[IT-1729] A lambda to stop or terminate all EC2 instances in an account

ConsoleCatzirl opened the pull request:
A lambda to stop or terminate all EC2 instances in an account.

ConsoleCatzirl committed:
[IT-1729] A lambda to stop or terminate all EC2 instances in an account

ConsoleCatzirl commented:
The error uploading to coveralls appears to be https://github.com/coverallsapp/coverage-reporter/issues/119

zaro0508 commented on README.md:
```diff
@@ -1,26 +1,44 @@
+
+### Running
+
+This lambda is triggered by a scheduled CloudWatch event at 2am UTC (8pm PST).
```
looks like this default can be overriden?  can you please add some documentation informing the user and give them info on how to override it?

Also what happens if this runs when when there is a packer build going or when some developer is running an AMI build in their testing workflow?

ConsoleCatzirl replied:
hm, good point, I should probably skip instances less than an hour old (should be more than enough),

zaro0508 commented on template.yaml:
```diff
@@ -1,41 +1,59 @@
+      PolicyDocument:
+        Version: 2012-10-17
+        Statement:
+          - Sid: ManageVolumes
```
this description does not make sense

ConsoleCatzirl replied:
whoops, copypasta

zaro0508 commented on README.md:
```diff
@@ -1,26 +1,44 @@
-# lambda-template
-A GitHub template for quickly starting a new AWS lambda project.
+# lambda-ec2-terminator
```
this lambda is doing something a little more general than just terminating instances so this name seems misleading.  Would it make sense to change it to something like `lambda-ec2-state` or `lambda-ec2-state-setter`?

ConsoleCatzirl replied:
I guess if I add more use cases that could make sense, but I'm not sure how useful it would be for Sage.

zaro0508 commented on template.yaml:
```diff
@@ -1,41 +1,59 @@
+    Type: String
+    Description: "Action to take on running instances"
+    Default: "STOP"
+    AllowedValues:
+      - "STOP"
+      - "TERMINATE"
```
I don't think cloudformation can check for case sensitivity therefore i think  `stop` and `terminate` are also a valid inputs.  If that's true then the python code needs to handle case insensitivity?

ConsoleCatzirl replied:
done

zaro0508 replied:
am i missing something?  I don't see any updates to handle case insensitivity.

ConsoleCatzirl replied:
The change I made was calling `.lower()` inside `set_instance_state` so that casing is ignored at the time of processing.

zaro0508 replied:
ohh i see it now.  i thought it would be done in `lambda_handler` right after reading in `EC2_ACTION` so that it's consistent throughout the app.

zaro0508 commented on ec2_terminator/app.py:
```diff
@@ -0,0 +1,157 @@
+            if ec2_instances:
+                found = True
+                if ec2_action == 'terminate':
+                    logging.info(f"Terminating Instances: {ec2_instances}")
```
should this be at debug level since there's already logging at info level below with `logging.info(message)`?

ConsoleCatzirl replied:
done

zaro0508 commented on ec2_terminator/app.py:
```diff
@@ -0,0 +1,157 @@
+    return instances
+
+
+def stop_instances(instances, region):
```
couldn't this function and `terminate_instances` be combined into a general `set_instance_state(instances, state, region)` function?  Where the state is their `stop` or `terminate`?

ConsoleCatzirl replied:
I took a stab at implementing this, but there isn't a similar boto3 function, so there's still two blocks of code for the two different cases of calling `client.stop_instances()` or `client.terminate_instances` and then parsing the output for either `StoppingInstances` or `TerminatingInstances`. This does make it easier to add other cases, but the only other similar boto3 functions are `client.start_instances`, `client.reboot_instances`, `client.monitor_instances`, and `client.unmonitor_instances`. I guess I can add them to make this more of a general lifecycle manager, what are your thoughts on the usefulness of those functions?

zaro0508 replied:
the other lifecycle options are probably not useful for us currently however setting it up to allow us to easily add them is useful.  I'm thinking either someone else will find them useful or that AWS will add additional options that will be useful for other use cases.

ConsoleCatzirl committed:
Add a tag for ignoring specific instances

Also fix logging in the lambda runtime environment.

ConsoleCatzirl committed:
combine stop_instances and terminate_instances into set_instance_state

ConsoleCatzirl committed:
review feedback

ConsoleCatzirl committed:
fix case insensitivity for EC2_ACTION env var

remove double-processing of EC2_ACTION env var and only rely on
processing the value in set_instance_state.
